File: model/src/main/java/net/qutester/model/infrastructure/path/Path.kt
package net.qutester.model.infrastructure.path

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import java.nio.file.Paths

data class Path @JsonCreator constructor(
        @JsonProperty("directories") val directories: List<String>,
        @JsonProperty("fileName") val fileName: String?,
        @JsonProperty("fileExtension") val fileExtension: String?) {

    companion object {
        val EMPTY = Path(emptyList(), null, null)

        fun createInstance(pathAsString: String): Path {

            var pathsPart = pathAsString.split("/", "\\")
            var fileName:String? = null;
            var extension:String? = null;

            val lastPathPart = pathsPart.last().toString()
            if (lastPathPart.contains('.')) {
                pathsPart = pathsPart.dropLast(1)
                fileName = lastPathPart.substringBefore(".")
                extension = lastPathPart.substringAfter(".")
            }

            return Path(pathsPart.map { it.toString() }, fileName, extension)
        }
    }

    @JsonIgnore
    fun toJavaPath(): java.nio.file.Path {
        return Paths.get(this.toString())
    }

    override fun toString(): String {
        var result = "";
        for ((index, directory) in directories.withIndex()) {
            result += directory
            if (index < directories.size - 1) {
                result += "/"
            }
        }

        if (fileName != null) {
            if (result.isNotEmpty()) {
                result += "/"
            }
            result += fileName
        }
        if (fileExtension != null) {
            result += "." + fileExtension

        }
        return result
    }

    @JsonIgnore
    fun isFile(): Boolean {
        return fileName != null || fileExtension != null
    }

    @JsonIgnore
    fun isDirectory(): Boolean {
        return !isFile()
    }

    @JsonIgnore
    fun resolve(path: Path): Path {
        if (this.isFile()) {
            throw RuntimeException("Can't resolve relative path to a file")
        }

        return Path(this.directories + path.directories, path.fileName, path.fileExtension)
    }

    @JsonIgnore
    fun isEmpty(): Boolean {
        return directories.isEmpty() && fileName == null && fileExtension == null;
    }
}

fun java.nio.file.Path.toMyPath(): Path {
    return Path.createInstance(this.toString())
}

File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/report-model/model/path.ts
export class Path {

    constructor(public readonly directories: Array<string>,
                public readonly fileName: string,
                public readonly fileExtension: string) {}

    static createInstance(pathAsString: string): Path {
        if (!pathAsString) {
            return Path.createInstanceOfEmptyPath();
        }

        let pathsPart = pathAsString.split("/").filter(value => !Path.stringIsEmpty(value));

        let fileName:string = null;
        let extension:string = null;

        let dotIndexInFileName = 0;
        if (pathsPart.length > 0) {
            let lastPathPart = pathsPart[pathsPart.length - 1];
            dotIndexInFileName = lastPathPart.search('\\.');

            if (dotIndexInFileName > 0) {
                pathsPart.splice(pathsPart.length - 1, 1);
                fileName = lastPathPart.substring(0, dotIndexInFileName);
                extension = lastPathPart.substring(dotIndexInFileName + 1, lastPathPart.length)
            }
        }

        return new Path(pathsPart, fileName, extension)
    }

    static createInstanceOfEmptyPath() {
        return new Path([], null, null)
    }

    private static stringIsEmpty(str: string): boolean {
        if (!str) {
            return true;
        }

        if (str.trim && str.trim() === "") {
            return true;
        }

        return false;
    }

    toDirectoryString(): string {
        let result: string = this.directories.join('/');

        if (!result.startsWith('/')) {
            result = '/' + result;
        }
        if (!result.endsWith('/')) {
            result = result + '/';
        }

        return result
    }

    toString(): string {
        let result = this.toDirectoryString();

        if (this.fileName) {
            result += this.fileName
        }
        if (this.fileExtension != null) {
            result += "." + this.fileExtension
        }

        return result
    }

}
